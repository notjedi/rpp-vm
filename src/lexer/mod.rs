mod tokens;
use itertools::Itertools;
pub(crate) use tokens::{KeyWord, Literal, Token};

use std::{i64, iter::Peekable, str::Chars};
use thiserror::Error;

#[derive(Debug, Error)]
#[allow(dead_code)]
pub(crate) enum LexError {
    #[error("unexpected character {0}")]
    UnexpectedChar(char),
    // TODO: `expected` can be multiple characters
    #[error("expected: {expected}, found: {found}")]
    MissingExpectedChar { expected: char, found: char },
}

pub(crate) struct Lexer<'a> {
    input: Peekable<Chars<'a>>,
}

impl<'a> Lexer<'a> {
    pub(crate) fn new(input: &'a str) -> Self {
        Self {
            input: input.chars().peekable(),
        }
    }

    pub(crate) fn tokenize(program: &str) -> Result<Vec<Token>, LexError> {
        let mut lexer = Lexer::new(program);
        let mut tokens = Vec::new();
        loop {
            let token = lexer.advance_token()?;
            if token == Token::Eof {
                break;
            }
            tokens.push(token);
        }
        Ok(tokens)
    }

    #[inline]
    fn consume(&mut self) -> Option<char> {
        self.input.next()
    }

    #[inline]
    fn peek(&mut self) -> Option<char> {
        self.input.peek().copied()
    }

    #[inline]
    fn take_while(&mut self, predicate: impl Fn(&char) -> bool) -> String {
        self.input.peeking_take_while(predicate).collect::<String>()
    }

    #[inline]
    fn skip_whitespace(&mut self) {
        self.take_while(|&ch| ch.is_ascii_whitespace());
    }

    fn eat_number(&mut self) -> Token {
        // TODO: mare clean way to do this?
        let mut len = 0;
        let mut has_dot = false;
        let mut num_chars = [0 as char; 128];

        while let Some(ch) = self.peek() {
            if ch.is_ascii_digit() {
                self.consume();
            } else if ch == '.' && !has_dot {
                self.consume();
                has_dot = true;
            } else if ch == '_' {
                // NOTE: _ b/w numbers is not in the grammar afaik but we'll add it
                continue;
            } else {
                break;
            }
            num_chars[len] = ch;
            len += 1;
        }

        let num_str = num_chars[..len].iter().collect::<String>();
        // SAFETY: it's safe to unwrap here because we've made sure that the array only has digits and at max of one `.`
        if has_dot {
            let num = num_str.parse::<f64>().unwrap();
            Token::Literal(tokens::Literal::Float(num))
        } else {
            let num = num_str.parse::<i64>().unwrap();
            Token::Literal(tokens::Literal::Int(num))
        }
    }

    fn eat_literal_str(&mut self) -> Result<String, LexError> {
        // NOTE: does not support `"` inside the string
        self.consume();
        let line = self.take_while(|&ch| ch != '"');
        match self.peek() {
            Some('"') => {
                self.consume();
                Ok(line)
            }
            ch => Err(LexError::MissingExpectedChar {
                expected: '"',
                found: ch.unwrap_or_default(),
            }),
        }
    }

    fn eat_literal_char(&mut self) -> Result<char, LexError> {
        self.consume();
        let ch = self
            .consume()
            .ok_or(LexError::UnexpectedChar(char::default()));
        match self.peek() {
            Some('\'') => {
                self.consume();
                ch
            }
            peek_ch => Err(LexError::MissingExpectedChar {
                expected: '\'',
                found: peek_ch.unwrap_or_default(),
            }),
        }
    }

    #[inline]
    fn eat_line(&mut self) -> String {
        self.take_while(|&ch| ch == '\n')
    }

    #[inline]
    fn eat_ident(&mut self) -> String {
        self.take_while(|&ch| ch.is_ascii_alphabetic() || ch == '_')
    }

    #[inline]
    fn eat_n_idents(&mut self, n: u32) {
        (0..n).for_each(|_| {
            self.eat_ident();
        })
    }

    fn eat_potential_double_char_op(
        &mut self,
        expected: char,
        one_char_kind: Token,
        double_char_kind: Token,
    ) -> Token {
        if let Some(next) = self.peek() {
            if next == expected {
                self.consume();
                return double_char_kind;
            }
            return one_char_kind;
        }
        one_char_kind
    }

    fn eat_potential_double_char_or_err(
        &mut self,
        expected: char,
        token: Token,
    ) -> Result<Token, LexError> {
        if let Some(found) = self.peek() {
            if found == expected {
                self.consume();
                return Ok(token);
            }
            return Err(LexError::MissingExpectedChar { expected, found });
        }
        Err(LexError::UnexpectedChar(char::default()))
    }

    fn eat_punctuation(&mut self) -> Result<Token, LexError> {
        let ch = self.consume().unwrap();

        let token = match ch {
            '+' => KeyWord::Sum.into(),
            '-' => match self.peek() {
                Some('0'..='9') => match self.eat_number() {
                    Token::Literal(Literal::Float(float_num)) => {
                        Token::Literal(Literal::Float(-float_num))
                    }
                    Token::Literal(Literal::Int(int_num)) => Token::Literal(Literal::Int(-int_num)),
                    _ => unreachable!(),
                },
                _ => KeyWord::Sub.into(),
            },
            '*' => KeyWord::Mul.into(),
            '/' => KeyWord::Div.into(),
            '%' => KeyWord::Mod.into(),

            ';' => KeyWord::SemiColon.into(),
            '{' => KeyWord::LeftBrace.into(),
            '}' => KeyWord::RightBrace.into(),

            '<' => self.eat_potential_double_char_op(
                '=',
                KeyWord::LessThan.into(),
                KeyWord::LessThanEqual.into(),
            ),
            '>' => self.eat_potential_double_char_op(
                '=',
                KeyWord::GreaterThan.into(),
                KeyWord::GreaterThanEqual.into(),
            ),

            '=' => self.eat_potential_double_char_or_err('=', KeyWord::Equal.into())?,
            ':' => self.eat_potential_double_char_or_err('=', KeyWord::DeclareAlt.into())?,
            '!' => match self.eat_potential_double_char_or_err('=', KeyWord::NotEqual.into()) {
                Ok(token) => token,
                _ => match self.peek() {
                    Some('!') => {
                        self.consume();
                        Token::Comment(self.eat_line())
                    }
                    ch => return Err(LexError::UnexpectedChar(ch.unwrap_or_default())),
                },
            },

            ch => return Err(LexError::UnexpectedChar(ch)),
        };
        Ok(token)
    }

    fn peek_n_words(&self, n: u32) -> String {
        let mut num_spaces = 0;
        self.input
            .clone()
            .take_while(|&ch| {
                if ch == ' ' {
                    num_spaces += 1;
                }
                (ch.is_ascii_alphabetic() || ch == '_' || ch == ' ') && num_spaces <= (n - 1)
            })
            .collect::<String>()
    }

    fn match_potential_single_word_kw(&self) -> Option<KeyWord> {
        match self.peek_n_words(1).as_str() {
            const { KeyWord::Print.as_str() } => KeyWord::Print.into(),
            const { KeyWord::EndFunc.as_str() } => KeyWord::EndFunc.into(),
            const { KeyWord::ForStart.as_str() } => KeyWord::ForStart.into(),
            const { KeyWord::BoolTrue.as_str() } => KeyWord::BoolTrue.into(),
            const { KeyWord::BoolFalse.as_str() } => KeyWord::BoolFalse.into(),
            const { KeyWord::ProgramEnd.as_str() } => KeyWord::ProgramEnd.into(),
            _ => None,
        }
    }

    fn match_potential_double_word_kw(&self) -> Option<KeyWord> {
        match self.peek_n_words(2).as_str() {
            const { KeyWord::Assign.as_str() } => KeyWord::Assign.into(),
            const { KeyWord::Declare.as_str() } => KeyWord::Declare.into(),
            const { KeyWord::EndBlock.as_str() } => KeyWord::EndBlock.into(),
            const { KeyWord::FuncCall.as_str() } => KeyWord::FuncCall.into(),
            const { KeyWord::BreakLoop.as_str() } => KeyWord::BreakLoop.into(),
            const { KeyWord::ProgramStart.as_str() } => KeyWord::ProgramStart.into(),
            const { KeyWord::StartDeclare.as_str() } => KeyWord::StartDeclare.into(),
            const { KeyWord::ForRangeStart.as_str() } => KeyWord::ForRangeStart.into(),
            _ => None,
        }
    }

    fn match_potential_triple_word_kw(&self) -> Option<KeyWord> {
        match self.peek_n_words(3).as_str() {
            const { KeyWord::IfCond.as_str() } => KeyWord::IfCond.into(),
            const { KeyWord::WhileLoop.as_str() } => KeyWord::WhileLoop.into(),
            const { KeyWord::FuncReturn.as_str() } => KeyWord::FuncReturn.into(),
            const { KeyWord::ForRangeEnd.as_str() } => KeyWord::ForRangeEnd.into(),
            _ => None,
        }
    }

    fn match_potential_four_word_kw(&self) -> Option<KeyWord> {
        match self.peek_n_words(4).as_str() {
            const { KeyWord::ElseCond.as_str() } => KeyWord::ElseCond.into(),
            const { KeyWord::FuncDeclare.as_str() } => KeyWord::FuncDeclare.into(),
            _ => None,
        }
    }

    fn match_keyword(&mut self) -> Token {
        let kw = None
            .or_else(|| {
                self.match_potential_four_word_kw().and_then(|kw| {
                    self.eat_n_idents(4);
                    Some(kw)
                })
            })
            .or_else(|| {
                self.match_potential_triple_word_kw().and_then(|kw| {
                    self.eat_n_idents(3);
                    Some(kw)
                })
            })
            .or_else(|| {
                self.match_potential_double_word_kw().and_then(|kw| {
                    self.eat_n_idents(2);
                    Some(kw)
                })
            })
            .or_else(|| {
                self.match_potential_single_word_kw().and_then(|kw| {
                    self.eat_ident();
                    Some(kw)
                })
            });
        kw.map_or_else(|| Token::Ident(self.eat_ident()), |kw| Token::KeyWord(kw))
    }

    pub(crate) fn advance_token(&mut self) -> Result<Token, LexError> {
        self.skip_whitespace();

        let ch = match self.peek() {
            Some(ch) => ch,
            None => return Ok(Token::Eof),
        };

        let token = match ch {
            '"' => Token::Literal(Literal::Str(self.eat_literal_str()?)),
            '\'' => Token::Literal(Literal::Char(self.eat_literal_char()?)),
            '0'..='9' => self.eat_number(),
            'a'..='z' | 'A'..='Z' | '_' => self.match_keyword(),
            _ => self.eat_punctuation()?,
        };

        Ok(token)
    }
}
