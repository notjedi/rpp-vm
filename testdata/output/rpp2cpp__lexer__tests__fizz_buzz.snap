---
source: src/lexer/mod.rs
expression: snapshot_string
---
EN VAZHI THANI VAZHI myfunc_one
^^^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(FuncDeclare), span: Span { row: 0, col: 0, length: 20 } }
                     ^^^^^^^^^^ Token { kind: Ident("myfunc_one"), span: Span { row: 0, col: 21, length: 10 } }
    AANDAVAN SOLLRAN ix ARUNACHALAM SEIYARAN 100;
    ^^^^^^^^^^^^^^^^ Token { kind: KeyWord(StartDeclare), span: Span { row: 1, col: 4, length: 16 } }
                     ^^ Token { kind: Ident("ix"), span: Span { row: 1, col: 21, length: 2 } }
                        ^^^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(Declare), span: Span { row: 1, col: 24, length: 20 } }
                                             ^^^ Token { kind: Literal(Int(100)), span: Span { row: 1, col: 45, length: 3 } }
                                                ^ Token { kind: KeyWord(SemiColon), span: Span { row: 1, col: 48, length: 1 } }
    DOT "returning ix =" ix "to main";
    ^^^ Token { kind: KeyWord(Print), span: Span { row: 2, col: 4, length: 3 } }
        ^^^^^^^^^^^^^^^^ Token { kind: Literal(Str("returning ix =")), span: Span { row: 2, col: 8, length: 16 } }
                         ^^ Token { kind: Ident("ix"), span: Span { row: 2, col: 25, length: 2 } }
                            ^^^^^^^^^ Token { kind: Literal(Str("to main")), span: Span { row: 2, col: 28, length: 9 } }
                                     ^ Token { kind: KeyWord(SemiColon), span: Span { row: 2, col: 37, length: 1 } }
    IDHU EPDI IRUKKU ix;
    ^^^^^^^^^^^^^^^^ Token { kind: KeyWord(FuncReturn), span: Span { row: 3, col: 4, length: 16 } }
                     ^^ Token { kind: Ident("ix"), span: Span { row: 3, col: 21, length: 2 } }
                       ^ Token { kind: KeyWord(SemiColon), span: Span { row: 3, col: 23, length: 1 } }
MARAKKADHINGA
^^^^^^^^^^^^^ Token { kind: KeyWord(EndFunc), span: Span { row: 4, col: 0, length: 13 } }

LAKSHMI START
^^^^^^^^^^^^^ Token { kind: KeyWord(ProgramStart), span: Span { row: 6, col: 0, length: 13 } }
    !! checking exprs
    ^^^^^^^^^^^^^^^^^ Token { kind: Comment(" checking exprs"), span: Span { row: 7, col: 4, length: 17 } }
    25 + 15;
    ^^ Token { kind: Literal(Int(25)), span: Span { row: 8, col: 4, length: 2 } }
       ^ Token { kind: KeyWord(Sum), span: Span { row: 8, col: 7, length: 1 } }
         ^^ Token { kind: Literal(Int(15)), span: Span { row: 8, col: 9, length: 2 } }
           ^ Token { kind: KeyWord(SemiColon), span: Span { row: 8, col: 11, length: 1 } }
    25 - 15;
    ^^ Token { kind: Literal(Int(25)), span: Span { row: 9, col: 4, length: 2 } }
       ^ Token { kind: KeyWord(Sub), span: Span { row: 9, col: 7, length: 1 } }
         ^^ Token { kind: Literal(Int(15)), span: Span { row: 9, col: 9, length: 2 } }
           ^ Token { kind: KeyWord(SemiColon), span: Span { row: 9, col: 11, length: 1 } }
    5.5 * -5;
    ^^^ Token { kind: Literal(Float(5.5)), span: Span { row: 10, col: 4, length: 3 } }
        ^ Token { kind: KeyWord(Mul), span: Span { row: 10, col: 8, length: 1 } }
          ^ Token { kind: KeyWord(Sub), span: Span { row: 10, col: 10, length: 1 } }
           ^ Token { kind: Literal(Int(5)), span: Span { row: 10, col: 11, length: 1 } }
            ^ Token { kind: KeyWord(SemiColon), span: Span { row: 10, col: 12, length: 1 } }
    5 / 5;
    ^ Token { kind: Literal(Int(5)), span: Span { row: 11, col: 4, length: 1 } }
      ^ Token { kind: KeyWord(Div), span: Span { row: 11, col: 6, length: 1 } }
        ^ Token { kind: Literal(Int(5)), span: Span { row: 11, col: 8, length: 1 } }
         ^ Token { kind: KeyWord(SemiColon), span: Span { row: 11, col: 9, length: 1 } }
    51 % 5;
    ^^ Token { kind: Literal(Int(51)), span: Span { row: 12, col: 4, length: 2 } }
       ^ Token { kind: KeyWord(Mod), span: Span { row: 12, col: 7, length: 1 } }
         ^ Token { kind: Literal(Int(5)), span: Span { row: 12, col: 9, length: 1 } }
          ^ Token { kind: KeyWord(SemiColon), span: Span { row: 12, col: 10, length: 1 } }

    !! testing while loop
    ^^^^^^^^^^^^^^^^^^^^^ Token { kind: Comment(" testing while loop"), span: Span { row: 14, col: 4, length: 21 } }
    BABA COUNTING STARTS True {
    ^^^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(WhileLoop), span: Span { row: 15, col: 4, length: 20 } }
                         ^^^^ Token { kind: Literal(BoolTrue), span: Span { row: 15, col: 25, length: 4 } }
                              ^ Token { kind: KeyWord(LeftBrace), span: Span { row: 15, col: 30, length: 1 } }
        DOT ix;
        ^^^ Token { kind: KeyWord(Print), span: Span { row: 16, col: 8, length: 3 } }
            ^^ Token { kind: Ident("ix"), span: Span { row: 16, col: 12, length: 2 } }
              ^ Token { kind: KeyWord(SemiColon), span: Span { row: 16, col: 14, length: 1 } }
        ix BHAJJI SAAPDU ix + 1;
        ^^ Token { kind: Ident("ix"), span: Span { row: 17, col: 8, length: 2 } }
           ^^^^^^^^^^^^^ Token { kind: KeyWord(Assign), span: Span { row: 17, col: 11, length: 13 } }
                         ^^ Token { kind: Ident("ix"), span: Span { row: 17, col: 25, length: 2 } }
                            ^ Token { kind: KeyWord(Sum), span: Span { row: 17, col: 28, length: 1 } }
                              ^ Token { kind: Literal(Int(1)), span: Span { row: 17, col: 30, length: 1 } }
                               ^ Token { kind: KeyWord(SemiColon), span: Span { row: 17, col: 31, length: 1 } }
        EN PEAR MANICKAM ix >= 5 {
        ^^^^^^^^^^^^^^^^ Token { kind: KeyWord(IfCond), span: Span { row: 18, col: 8, length: 16 } }
                         ^^ Token { kind: Ident("ix"), span: Span { row: 18, col: 25, length: 2 } }
                            ^^ Token { kind: KeyWord(GreaterThanEqual), span: Span { row: 18, col: 28, length: 2 } }
                               ^ Token { kind: Literal(Int(5)), span: Span { row: 18, col: 31, length: 1 } }
                                 ^ Token { kind: KeyWord(LeftBrace), span: Span { row: 18, col: 33, length: 1 } }
            DOT "breaking out of loop...";
            ^^^ Token { kind: KeyWord(Print), span: Span { row: 19, col: 12, length: 3 } }
                ^^^^^^^^^^^^^^^^^^^^^^^^^ Token { kind: Literal(Str("breaking out of loop...")), span: Span { row: 19, col: 16, length: 25 } }
                                         ^ Token { kind: KeyWord(SemiColon), span: Span { row: 19, col: 41, length: 1 } }
            BLACK SHEEP;
            ^^^^^^^^^^^ Token { kind: KeyWord(BreakLoop), span: Span { row: 20, col: 12, length: 11 } }
                       ^ Token { kind: KeyWord(SemiColon), span: Span { row: 20, col: 23, length: 1 } }
        } KATHAM KATHAM;
        ^ Token { kind: KeyWord(RightBrace), span: Span { row: 21, col: 8, length: 1 } }
          ^^^^^^^^^^^^^ Token { kind: KeyWord(EndBlock), span: Span { row: 21, col: 10, length: 13 } }
                       ^ Token { kind: KeyWord(SemiColon), span: Span { row: 21, col: 23, length: 1 } }
    } KATHAM KATHAM;
    ^ Token { kind: KeyWord(RightBrace), span: Span { row: 22, col: 4, length: 1 } }
      ^^^^^^^^^^^^^ Token { kind: KeyWord(EndBlock), span: Span { row: 22, col: 6, length: 13 } }
                   ^ Token { kind: KeyWord(SemiColon), span: Span { row: 22, col: 19, length: 1 } }

    y CHUMMA ADHURUDHULA myfunc_one;
    ^ Token { kind: Ident("y"), span: Span { row: 24, col: 4, length: 1 } }
      ^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(FuncCall), span: Span { row: 24, col: 6, length: 18 } }
                         ^^^^^^^^^^ Token { kind: Ident("myfunc_one"), span: Span { row: 24, col: 25, length: 10 } }
                                   ^ Token { kind: KeyWord(SemiColon), span: Span { row: 24, col: 35, length: 1 } }

    AANDAVAN SOLLRAN ix ARUNACHALAM SEIYARAN 1;
    ^^^^^^^^^^^^^^^^ Token { kind: KeyWord(StartDeclare), span: Span { row: 26, col: 4, length: 16 } }
                     ^^ Token { kind: Ident("ix"), span: Span { row: 26, col: 21, length: 2 } }
                        ^^^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(Declare), span: Span { row: 26, col: 24, length: 20 } }
                                             ^ Token { kind: Literal(Int(1)), span: Span { row: 26, col: 45, length: 1 } }
                                              ^ Token { kind: KeyWord(SemiColon), span: Span { row: 26, col: 46, length: 1 } }
    AANDAVAN SOLLRAN range ARUNACHALAM SEIYARAN 16;
    ^^^^^^^^^^^^^^^^ Token { kind: KeyWord(StartDeclare), span: Span { row: 27, col: 4, length: 16 } }
                     ^^^^^ Token { kind: Ident("range"), span: Span { row: 27, col: 21, length: 5 } }
                           ^^^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(Declare), span: Span { row: 27, col: 27, length: 20 } }
                                                ^^ Token { kind: Literal(Int(16)), span: Span { row: 27, col: 48, length: 2 } }
                                                  ^ Token { kind: KeyWord(SemiColon), span: Span { row: 27, col: 50, length: 1 } }

    NAA 1 THADAVA SONNA range THADAVA SONNA MADHRI {
    ^^^ Token { kind: KeyWord(ForStart), span: Span { row: 29, col: 4, length: 3 } }
        ^ Token { kind: Literal(Int(1)), span: Span { row: 29, col: 8, length: 1 } }
          ^^^^^^^^^^^^^ Token { kind: KeyWord(ForRangeStart), span: Span { row: 29, col: 10, length: 13 } }
                        ^^^^^ Token { kind: Ident("range"), span: Span { row: 29, col: 24, length: 5 } }
                              ^^^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(ForRangeEnd), span: Span { row: 29, col: 30, length: 20 } }
                                                   ^ Token { kind: KeyWord(LeftBrace), span: Span { row: 29, col: 51, length: 1 } }
        EN PEAR MANICKAM ix%15==0 {
        ^^^^^^^^^^^^^^^^ Token { kind: KeyWord(IfCond), span: Span { row: 30, col: 8, length: 16 } }
                         ^^ Token { kind: Ident("ix"), span: Span { row: 30, col: 25, length: 2 } }
                           ^ Token { kind: KeyWord(Mod), span: Span { row: 30, col: 27, length: 1 } }
                            ^^ Token { kind: Literal(Int(15)), span: Span { row: 30, col: 28, length: 2 } }
                              ^^ Token { kind: KeyWord(Equal), span: Span { row: 30, col: 30, length: 2 } }
                                ^ Token { kind: Literal(Int(0)), span: Span { row: 30, col: 32, length: 1 } }
                                  ^ Token { kind: KeyWord(LeftBrace), span: Span { row: 30, col: 34, length: 1 } }
            DOT "FizzBuzz";
            ^^^ Token { kind: KeyWord(Print), span: Span { row: 31, col: 12, length: 3 } }
                ^^^^^^^^^^ Token { kind: Literal(Str("FizzBuzz")), span: Span { row: 31, col: 16, length: 10 } }
                          ^ Token { kind: KeyWord(SemiColon), span: Span { row: 31, col: 26, length: 1 } }
        } ENAKKU INNURU PEAR IRUKKU {
        ^ Token { kind: KeyWord(RightBrace), span: Span { row: 32, col: 8, length: 1 } }
          ^^^^^^^^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(ElseCond), span: Span { row: 32, col: 10, length: 25 } }
                                    ^ Token { kind: KeyWord(LeftBrace), span: Span { row: 32, col: 36, length: 1 } }
            EN PEAR MANICKAM ix%3==0 {
            ^^^^^^^^^^^^^^^^ Token { kind: KeyWord(IfCond), span: Span { row: 33, col: 12, length: 16 } }
                             ^^ Token { kind: Ident("ix"), span: Span { row: 33, col: 29, length: 2 } }
                               ^ Token { kind: KeyWord(Mod), span: Span { row: 33, col: 31, length: 1 } }
                                ^ Token { kind: Literal(Int(3)), span: Span { row: 33, col: 32, length: 1 } }
                                 ^^ Token { kind: KeyWord(Equal), span: Span { row: 33, col: 33, length: 2 } }
                                   ^ Token { kind: Literal(Int(0)), span: Span { row: 33, col: 35, length: 1 } }
                                     ^ Token { kind: KeyWord(LeftBrace), span: Span { row: 33, col: 37, length: 1 } }
                DOT "Fizz";
                ^^^ Token { kind: KeyWord(Print), span: Span { row: 34, col: 16, length: 3 } }
                    ^^^^^^ Token { kind: Literal(Str("Fizz")), span: Span { row: 34, col: 20, length: 6 } }
                          ^ Token { kind: KeyWord(SemiColon), span: Span { row: 34, col: 26, length: 1 } }
            } ENAKKU INNURU PEAR IRUKKU {
            ^ Token { kind: KeyWord(RightBrace), span: Span { row: 35, col: 12, length: 1 } }
              ^^^^^^^^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(ElseCond), span: Span { row: 35, col: 14, length: 25 } }
                                        ^ Token { kind: KeyWord(LeftBrace), span: Span { row: 35, col: 40, length: 1 } }
                EN PEAR MANICKAM ix%5==0 {
                ^^^^^^^^^^^^^^^^ Token { kind: KeyWord(IfCond), span: Span { row: 36, col: 16, length: 16 } }
                                 ^^ Token { kind: Ident("ix"), span: Span { row: 36, col: 33, length: 2 } }
                                   ^ Token { kind: KeyWord(Mod), span: Span { row: 36, col: 35, length: 1 } }
                                    ^ Token { kind: Literal(Int(5)), span: Span { row: 36, col: 36, length: 1 } }
                                     ^^ Token { kind: KeyWord(Equal), span: Span { row: 36, col: 37, length: 2 } }
                                       ^ Token { kind: Literal(Int(0)), span: Span { row: 36, col: 39, length: 1 } }
                                         ^ Token { kind: KeyWord(LeftBrace), span: Span { row: 36, col: 41, length: 1 } }
                    DOT "Buzz";
                    ^^^ Token { kind: KeyWord(Print), span: Span { row: 37, col: 20, length: 3 } }
                        ^^^^^^ Token { kind: Literal(Str("Buzz")), span: Span { row: 37, col: 24, length: 6 } }
                              ^ Token { kind: KeyWord(SemiColon), span: Span { row: 37, col: 30, length: 1 } }
                } ENAKKU INNURU PEAR IRUKKU {
                ^ Token { kind: KeyWord(RightBrace), span: Span { row: 38, col: 16, length: 1 } }
                  ^^^^^^^^^^^^^^^^^^^^^^^^^ Token { kind: KeyWord(ElseCond), span: Span { row: 38, col: 18, length: 25 } }
                                            ^ Token { kind: KeyWord(LeftBrace), span: Span { row: 38, col: 44, length: 1 } }
                    DOT ix;
                    ^^^ Token { kind: KeyWord(Print), span: Span { row: 39, col: 20, length: 3 } }
                        ^^ Token { kind: Ident("ix"), span: Span { row: 39, col: 24, length: 2 } }
                          ^ Token { kind: KeyWord(SemiColon), span: Span { row: 39, col: 26, length: 1 } }
                } KATHAM KATHAM;
                ^ Token { kind: KeyWord(RightBrace), span: Span { row: 40, col: 16, length: 1 } }
                  ^^^^^^^^^^^^^ Token { kind: KeyWord(EndBlock), span: Span { row: 40, col: 18, length: 13 } }
                               ^ Token { kind: KeyWord(SemiColon), span: Span { row: 40, col: 31, length: 1 } }
            } KATHAM KATHAM;
            ^ Token { kind: KeyWord(RightBrace), span: Span { row: 41, col: 12, length: 1 } }
              ^^^^^^^^^^^^^ Token { kind: KeyWord(EndBlock), span: Span { row: 41, col: 14, length: 13 } }
                           ^ Token { kind: KeyWord(SemiColon), span: Span { row: 41, col: 27, length: 1 } }
        } KATHAM KATHAM;
        ^ Token { kind: KeyWord(RightBrace), span: Span { row: 42, col: 8, length: 1 } }
          ^^^^^^^^^^^^^ Token { kind: KeyWord(EndBlock), span: Span { row: 42, col: 10, length: 13 } }
                       ^ Token { kind: KeyWord(SemiColon), span: Span { row: 42, col: 23, length: 1 } }
        ix BHAJJI SAAPDU ix+1;
        ^^ Token { kind: Ident("ix"), span: Span { row: 43, col: 8, length: 2 } }
           ^^^^^^^^^^^^^ Token { kind: KeyWord(Assign), span: Span { row: 43, col: 11, length: 13 } }
                         ^^ Token { kind: Ident("ix"), span: Span { row: 43, col: 25, length: 2 } }
                           ^ Token { kind: KeyWord(Sum), span: Span { row: 43, col: 27, length: 1 } }
                            ^ Token { kind: Literal(Int(1)), span: Span { row: 43, col: 28, length: 1 } }
                             ^ Token { kind: KeyWord(SemiColon), span: Span { row: 43, col: 29, length: 1 } }
    } KATHAM KATHAM;
    ^ Token { kind: KeyWord(RightBrace), span: Span { row: 44, col: 4, length: 1 } }
      ^^^^^^^^^^^^^ Token { kind: KeyWord(EndBlock), span: Span { row: 44, col: 6, length: 13 } }
                   ^ Token { kind: KeyWord(SemiColon), span: Span { row: 44, col: 19, length: 1 } }
MAGIZHCHI
^^^^^^^^^ Token { kind: KeyWord(ProgramEnd), span: Span { row: 45, col: 0, length: 9 } }

